content/blog/2019-09-25-vril-vector-bril/index.md

```diff
@@ -0,0 +1,137 @@
+Vril includes a new type `array` to the language. The definition of an array in Vril is a sequence of elements of any type, which includes arrays (no type checking carried out) and the length being any Bril literal. 
+
+### Initializing arrays in Vril
+Vril extends Bril to support arrays by adding an `init` operator to initialize an array of length `l` and set its elements to 0 as follows:
+```
+array_name : array = init l
+```
+
+Vril adds array operations to Bril of the form,
+```
+variable: type = aop args
+```
+### Supported array operations
+Vril can operate on arrays in two ways: 1) We can perform *scalar* accesses to arrays by moving one element of the array to a variable and performing any arithmetic or logic operation that are already supported in the original Bril, and 2) we can perform *vector* accesses to the arrays by using new vector operations that take arrays as arguments directly.
+
+In order to move data from an element of an array `arr[idx]` to a variable `var`, we can use `a2v` op:
+```
+var: int = a2v array_name idx
+```
+Similarly, we can move a value stored in a variable `var` into one of the array's element `arr[idx]`:
+```
+array_name: int = v2a var index
+```
+
+Constants are inserted to arrays through variables, just like the original Bril.
+
+Vector ops
+----------------------------
+
+### Vector ops in Vril
+There are two types of vector ops in Vril: configuration ops and arithmetic/logic ops. Vector arithmetic or logic operations take arrays as arguments and produce an array or a scalar value as a result. Configuration ops are used to adjust the amount of data parallelism needed by the program to the resources that the micro-architecture offers. For example, in a for-loop that has no inter-iteration dependences we could potentially run all iterations in parallel if we had enough hardware resources. Otherwise, we would run it in groups of iterations in parallel that the hardware can handle.
+
+### Adding vector ops to Vril
+For now there is only one configuration op: `setvl`, which is based on the [RISC-V V vector extension](https://github.com/riscv/riscv-v-spec).
+```
+vl: int = setvl val
+```
+which requests `val` elements to the hardware and returns the `vl`, which is the actual number of elements that the hardware can support. The pseudo-code of this instruction is: 
+```
+vl := (val <= max_vl)? val:max_vl;
+```
+where `max_vl` is the maximum amount of lanes or array elements that the micro-architecture can process in parallel.
+
+We have also implemented only one arithmetic vector op `vadd` as a proof of concept:
+```
+arr3: array = vadd arr1 arr2 idx
+```
+Which takes two arrays as arguments and performs element-wise vector addition to `vl` elements starting at element `idx`, and stores it into a third array. For now, if `arr1`, `arr2`, and `arr3` have different lengths, as long as elements in indices `i:i+vl-1` are valid for the three arrays (two sources and destination), the operation is valid. 
```

Comment:
Got it! This makes much more sense now that there's an `idx` parameter to `vadd`. Thanks for clarifying!